hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,197 @@
+             * a sphere of radius R, 1(TRUE) if inside.
+             */
+            Scalar r2 = dot(pos,pos);
+            // exit immediately if no collision is found.
+            /*
+            To avoid division by zero error later on, we exit immediately. (no collision could have occurred
+            if the particle speed is equal to zero)
+            */
+            Scalar v2 = dot(vel,vel);
+            if (r2 <= R2 || v2 == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
+
+            // Find the point of contact when the particle is just leaving the surface
+            /*
+            * Calculating the time spent outside-bounds requires the knowledge of the point of contact on the boundary.
+            * The point of contact can be calculated using geometrical considerations.
+            *
+            * We know the following quantities,
+            *    1. r(t+del_t)      (the point outside the sphere)
+            *    2. v(t)            (previous velocity)
+            *
+            * Assuming 'r*' is the point of contact on the spherical shell and dt is the time particle travelled
+            * outside the boundary.
+            *
+            * r* = r(t+del_t) - dx* * v(t)/|v|
+            * and |r*|**2 = R**2
+            *
+            * Solving the above equations,
+            * dx* = (r.v^) - sqrt((r.v^)**2 - r2 + R2)
+            * consequently, dt = dx* / |v|
+            *
+            * --> dt = ((r.v) - sqrt((r.v)**2 - v2*(r2-R2)))/v2
+            */
+
+            Scalar rv = dot(pos,vel);
+            dt = (rv - fast::sqrt(rv*rv-v2*(r2-R2)))/v2;
+
+            // backtrack the particle for time dt to get to point of contact
+            pos -= vel*dt;
+
+            // update velocity according to boundary conditions
+            /*
+             * Let n^ be the normal unit vector ar the point of contact.
+             * Therefore,
+             * v_perp = (v.n^)n^        -->         ((r.v)/R2)r         [R2, since the particle is on the surface]
+             * v_para = v-v_perp
+             */
+            if (m_bc == boundary::no_slip)
+                {
+                /* no-slip requires reflection of the tangential components.
+                 * Radial component reflected since no penetration of the surface is necessary.
+                 * This results in just flipping of all the velocity components.
+                 */
+                vel = -vel;
+                }
+            else if (m_bc == boundary::slip)
+                {
+                // tangential component of the velocity is unchanged.
+                // Radial component reflected since no penetration of the surface is necessary.
+                /*
+                 * v' = -v_perp + v_para = v - 2*v_perp
+                */
+                const Scalar3 vperp = dot(vel,pos)*pos/R2;
```

Comment:
`m_R2` and writing like this might make it clearer where the vector comes from:
```suggestion
                const Scalar3 vperp = (dot(vel,pos)/m_R2)*pos;
```